fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/domain/SearchQueryRenderer.java

```diff
@@ -522,8 +537,45 @@ protected WhereFragment getTokenFilter(QueryParameter queryParm, String paramAli
             // precedence drama, we simply wrap everything in parens just to be safe
             where.leftParen();
 
-            if (Modifier.IN.equals(queryParm.getModifier()) || Modifier.NOT_IN.equals(queryParm.getModifier()) ||
-                    Modifier.ABOVE.equals(queryParm.getModifier()) || Modifier.BELOW.equals(queryParm.getModifier())) {
+            if (Modifier.IN.equals(queryParm.getModifier()) ||
+                    Modifier.NOT_IN.equals(queryParm.getModifier()) ||
+                    Modifier.ABOVE.equals(queryParm.getModifier()) ||
+                    Modifier.BELOW.equals(queryParm.getModifier())) {
+                // IN and NOT_IN can follow the same code path here because of our assumption that
+                // if there's any matching parameter values then we'll also find a matching common_token_value_id
+                //
+                // Assume we have a valueset vs with codes like s1|c1 and s2|c2 and we have Basic resources `a` and `b`.
+                // Now the query comes like Basic?p:not-in=vs
+                //
+                // Scenario 1: `a` has the value `s1|c1` for the coding targeted by parameter p.
+                //
+                //    WHERE NOT EXISTS(
+                //    SELECT 1
+                //    FROM Basic_RESOURCE_TOKEN_REFS AS P2
+                //    WHERE P2.LOGICAL_RESOURCE_ID = LR1.LOGICAL_RESOURCE_ID
+                //      AND P2.PARAMETER_NAME_ID = 20046
+                //      AND ((P2.COMMON_TOKEN_VALUE_ID IN (1)))))) AS LR
+                //
+                // the select subquery selects 1 for `a`
+                // the select subquery selects empty for `b`
+                // so as a result it selects `b` and not `a`
+                //
+                //
+                // Scenario 2: none of the resources have any of the values from the valueset
+                //
+                //    WHERE NOT EXISTS(
+                //    SELECT 1
+                //    FROM Basic_RESOURCE_TOKEN_REFS AS P2
+                //    WHERE P2.LOGICAL_RESOURCE_ID = LR1.LOGICAL_RESOURCE_ID
+                //      AND P2.PARAMETER_NAME_ID = 20046
+                //      AND ((P2.COMMON_TOKEN_VALUE_ID IN (-1)))))) AS LR
```

Comment:
Just a thought...if we only have -1 in the set here, it might be worth replacing that with NULL. This _may_ let the database optimize the query more, because it knows that NULL will never match anything and therefore it could short-circuit the join (if it's clever enough). With -1, it still has to go fishing in the index to find that value. May also impact cardinality estimation (hopefully in a positive way).

Reply:
Its actually a bind variable (I had simplified it for my comment).  I went ahead and changed the behavior so that we bind a null instead of a -1 (at least for the case of token params).
For reference params, we use BaseWhereAdapter.eq which takes primitive values, so in that case I'm still using the `-1`.

We should probably resolve on a consistent approach to these.  Is it better to use literals?  Is it better to use null?

Reply:
Comparisons with common_token_value_id shouldn't use bind variables - we always want to use literals to make sure we get the best plan. With PostgreSQL it's more interesting, because it definitely peeks more at the bind values when optimizing, but other databases may not. Anyway, our standard approach is to always use literals where it's a database id (where there is little chance the query will be reused).

As for OR vs IN list, the optimizer will usually automatically convert...but I tend to prefer an IN list because it's neater.

If you *only* have -1 - e.g. WHERE foo_id IN (-1) then we should rewrite that to be WHERE foo_id IN (NULL). That gives the optimizer some useful information.

Reply:
Our current BaseWhereAdapter doesn't seem to support the use of NULL literals (in either eq or in expressions), so unless we change that we need to pick whether to use literals or whether to use NULLs.  For now, I opted to use literals across these two methods now and we can use #2688 if we want to add support for replacing the -1 values with NULLs.

Reply:
For completeness, I just wanted to update here and say we should probably use `WHERE foo_id = NULL` for the case where there is a single value. That probably gives us the best chance that the optimizer will see it and short-circuit the join (because nothing ever is equal to NULL). Of course, if we know this we could probably rewrite the query ourselves, but that is complicated, so we are delegating responsibility to the DB layer, in the hope it will do the right thing.